text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+
+The batch creation+sorting phase for OTLP Arrow represents almost all the time spent. Serialization and deserialization
+times are
+almost zero due to the use of Flatbuffer by Apache Arrow (ser/deser without parsing/unpacking). Compression and
+decompression
+times are low due to the fact that the size of the Arrow OTLP message before compression is more than 7 times
+smaller than an OTLP message with the same content.
+
+> It should be possible to significantly optimize the creation of OTLP Arrow batches for contexts where the structure
+> of the logs to be sent is known in advance.
+
+### Traces
+
+No benchmark has been conducted for traces due to lack of a dataset containing traces with links and events.
+
+Please feel free to share a dataset with such characteristics to complete this document.
+
+Please check this [open issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13626) for more
+details on a solution to facilitate data sharing.
+
+## Appendix C - Parameter Tuning and Design Optimization
+
+This section describes the systematic approach used to optimize the Arrow OTLP design and its parameters. The approach
+is
+based on an optimization technique called "blackbox optimization" which allows to describe the system to be optimized as
+a black box with a set of input parameters and an output corresponding to the value of a function to be optimized
+(maximize or minimize). This technique is described in detail in the
+article [Optimize your applications using Google Vertex AI Vizier](https://cloud.google.com/blog/products/ai-machine-learning/optimize-your-applications-using-google-vertex-ai-vizier)
+.
+
+In this analysis, the function to optimize is the compression ratio for a set of parameters. We will focus on optimizing
+the protocol design for logs (the same approach has been applied for metrics). The parameters explored are:
+
+* compression algorithm (Zlib, Zstd, and Lz4)
+* serialization mode
+  * normalized: resources, instrumentation libraries, metrics, logs, traces, events, and links are mapped in a
+  dedicated
+  RecordBatch. A set of primary and secondary keys are used to recreate the relationships between these different
+  entities.
+  * denormalized: resource, instrumentation library, event, and link fields are replicated for every metrics, logs,
+  and traces.
+* dictionary configuration
+  * min_row_count: The creation of a dictionary will be performed only on columns with more than `min_row_count`
+  elements.
+  * max_card: The creation of a dictionary will be performed only on columns with a cardinality lower than `max_card`.
+  * max_card_ratio: The creation of a dictionary will only be performed on columns with a ratio `card` / `size` <= `max_card_ratio`.
+  * max_sorted: Maximum number of sorted dictionaries (based on cardinality/total_size and avg_data_length).
+* batch_size: Size of the batch before serialization and compression.
```

Comment:
It would be interesting to the results of optimization when we exclude this as an optimization parameter, since it is not always a parameter that protocol implementation have control over.

Reply:
We are interested in studying performance on real data. As it stands, all our experimental data is synthesized from data that was aggregated on an existing collection path, we don't record requests in their original batch sizes so it's difficult to do this study with synthetic data. We're setting up a real cluster w/ an arrow bridge to learn more.